--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -553,11 +_,20 @@
     private static final int MAX_PER_TICK = io.papermc.paper.configuration.GlobalConfiguration.get().misc.maxJoinsPerTick; // Paper - Buffer joins to world
     private static int joinAttemptsThisTick; // Paper - Buffer joins to world
     private static int currTick; // Paper - Buffer joins to world
+    private static int tickSecond; // Pempek - Max joins per second
     public void tick() {
         this.flushQueue();
         // Paper start - Buffer joins to world
         if (Connection.currTick != net.minecraft.server.MinecraftServer.currentTick) {
             Connection.currTick = net.minecraft.server.MinecraftServer.currentTick;
+            // Pempek start - Max joins per second
+            if (org.stegripe.pempek.PempekConfig.maxJoinsPerSecond) {
+                if (++Connection.tickSecond > 20) {
+                    Connection.tickSecond = 0;
+                    Connection.joinAttemptsThisTick = 0;
+                }
+            } else
+            // Pempek end - Max joins per second
             Connection.joinAttemptsThisTick = 0;
         }
         // Paper end - Buffer joins to world
