--- a/net/minecraft/world/item/BlockItem.java
+++ b/net/minecraft/world/item/BlockItem.java
@@ -134,7 +_,16 @@
     protected boolean updateCustomBlockEntityTag(
         final BlockPos pos, final Level level, final @Nullable Player player, final ItemStack itemStack, final BlockState placedState
     ) {
-        return updateCustomBlockEntityTag(level, player, pos, itemStack);
+        // Pempek start - Persistent BlockEntity Lore and DisplayName
+        boolean handled = updateCustomBlockEntityTag(level, player, pos, itemStack);
+        if (level.pempekConfig.persistentTileEntityLore) {
+            BlockEntity blockEntity1 = level.getBlockEntity(pos);
+            if (blockEntity1 != null) {
+                blockEntity1.setPersistentLore(itemStack.getOrDefault(DataComponents.LORE, net.minecraft.world.item.component.ItemLore.EMPTY));
+            }
+        }
+        return handled;
+        // Pempek end - Persistent BlockEntity Lore and DisplayName
     }
 
     protected @Nullable BlockState getPlacementState(final BlockPlaceContext context) {
@@ -197,6 +_,7 @@
                 }
 
                 if (!type.onlyOpCanSetNbt() || player != null && (player.canUseGameMasterBlocks() || (player.getAbilities().instabuild && player.getBukkitEntity().hasPermission("minecraft.nbt.place")))) { // Spigot - add permission
+                    if (!(level.pempekConfig.silkTouchEnabled && blockEntity instanceof net.minecraft.world.level.block.entity.SpawnerBlockEntity && player.getBukkitEntity().hasPermission("pempek.drop.spawners"))) // Pempek - Silk touch spawners
                     return customData.loadInto(blockEntity, level.registryAccess());
                 }
 
@@ -236,6 +_,7 @@
     public void onDestroyed(final ItemEntity entity) {
         ItemContainerContents container = entity.getItem().set(DataComponents.CONTAINER, ItemContainerContents.EMPTY);
         if (container != null) {
+            if (entity.level().pempekConfig.shulkerBoxItemDropContentsWhenDestroyed && this.getBlock() instanceof ShulkerBoxBlock) // Pempek - option to disable shulker box items from dropping contents when destroyed
             ItemUtils.onContainerDestroyed(entity, container.nonEmptyItemCopyStream());
         }
     }
