--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -268,11 +_,11 @@
             JfrCommand.register(this.dispatcher);
         }
 
-        if (SharedConstants.DEBUG_CHASE_COMMAND) {
+        if (org.stegripe.pempek.PempekConfig.registerMinecraftDisabledCommands || SharedConstants.DEBUG_CHASE_COMMAND) { // Pempek - register disabled minecraft commands
             ChaseCommand.register(this.dispatcher);
         }
 
-        if (SharedConstants.DEBUG_DEV_COMMANDS || SharedConstants.IS_RUNNING_IN_IDE) {
+        if (org.stegripe.pempek.PempekConfig.registerMinecraftDebugCommands || SharedConstants.DEBUG_DEV_COMMANDS || SharedConstants.IS_RUNNING_IN_IDE) {  // Pempek - register minecraft debug commands
             RaidCommand.register(this.dispatcher, context);
             DebugPathCommand.register(this.dispatcher);
             DebugMobSpawningCommand.register(this.dispatcher);
@@ -300,6 +_,14 @@
             StopCommand.register(this.dispatcher);
             TransferCommand.register(this.dispatcher);
             WhitelistCommand.register(this.dispatcher);
+            org.stegripe.pempek.command.CreditsCommand.register(this.dispatcher); // Pempek - Add credits command
+            org.stegripe.pempek.command.DemoCommand.register(this.dispatcher); // Pempek - Add demo command
+            org.stegripe.pempek.command.PingCommand.register(this.dispatcher); // Pempek - Add ping command
+            org.stegripe.pempek.command.UptimeCommand.register(this.dispatcher); // Pempek - Add uptime command
+            org.stegripe.pempek.command.TPSBarCommand.register(this.dispatcher); // Pempek - Implement TPSBar
+            org.stegripe.pempek.command.CompassCommand.register(this.dispatcher); // Pempek - Add compass command
+            org.stegripe.pempek.command.RamBarCommand.register(this.dispatcher); // Pempek - Add rambar command
+            org.stegripe.pempek.command.RamCommand.register(this.dispatcher); // Pempek - Add ram command
         }
 
         if (commandSelection.includeIntegrated) {
@@ -523,6 +_,7 @@
     private void runSync(ServerPlayer player, java.util.Collection<String> bukkit, RootCommandNode<CommandSourceStack> root) {
         // Paper end - Perf: Async command map building
         new com.destroystokyo.paper.event.brigadier.AsyncPlayerSendCommandsEvent<CommandSourceStack>(player.getBukkitEntity(), (RootCommandNode) root, true).callEvent(); // Paper - Brigadier API
+        if (org.bukkit.event.player.PlayerCommandSendEvent.getHandlerList().getRegisteredListeners().length > 0) { // Pempek - Skip events if there's no listeners
         org.bukkit.event.player.PlayerCommandSendEvent event = new org.bukkit.event.player.PlayerCommandSendEvent(player.getBukkitEntity(), new java.util.LinkedHashSet<>(bukkit));
         event.getPlayer().getServer().getPluginManager().callEvent(event);
 
@@ -533,6 +_,8 @@
             }
         }
         // CraftBukkit end
+        } // Pempek - Skip events if there's no listeners
+
         player.connection.send(new ClientboundCommandsPacket(root, COMMAND_NODE_INSPECTOR));
     }
 
